## Git

Git is a version control system that allow us to coordinate with others to make work done effictively. Git is created by **Linus in only 2 weeks**! Now is the most popular tool used cross world.


If there is one thing that should be learned before real work, that's **Git**!


But what is `version contraol`? 

Supporse that there are 2 developers that are working on a project called `auto-ml-cl` that aim to provide a tool that to build machine learning models without any feature engineering, the main file is called `automl.py` file, supporse the same day that these 2 developers are working on this same file but with different functions, they are both have a copy file in their local computer, after a hard work day that there have both finished some functions design, then they need a way to `merge` their work into one place, so that we could make it workable, right? But how to make it? 

The naive way is that 1th person give the file directly to 2th person and let 2th person to `copy` his work into his own, right? Yes it's way to do it. But if we have 10 person are working on this same file? It's hard to make it just by one person, right? That's **Git** comes in, there is one `central repository` that holds these projects files, and we could `clone` it into our local folder, so what we want on these files, and to store these files in our local, after we have finished, then we could just `commit` to tell git: I have finished, please keep it. We could do it many times we want, and git will remember every time what I have changed, we could even move back to original work. After we have finished full work, then we need to make it into central repository so others will see it and continue their work, that's `pull` to remote server that holds the central repository.


That's high level explain for **Git**, this tutorial will give hands-on tutorial for how it works! Get your hands dirty! Let's start!


### 1. Install Git


Linux:


If you are using linux then just with command: `sudo apt-get install git`, that's easy right? Yes it's, as git is created for **Linux** community!


Windows:

First thing is to download git from official website with windows version: [Git Download](https://git-scm.com/downloads). Then just install it like other software.

### 2. Check Git

After we have installed it, then we need to test it is workable, so just open a windows terminal with command: `CMD`, then in the terminal type with `git version` to check that we have already install it correctly!

You may be with different version, but that's OK, most daily work will use some most common command that Git provide, so if you do face any version specific problem, don't hesitate to **Google** it!

In [1]:
! git version

Active code page: 65001
git version 2.29.0.windows.1


### 3. Configure Git

The first time we should tell git who we are with command: `git config --global user.email "you@example.com"` and `git config --global user.name "Your Name"`, if you don't need want to make it to global, then remove `global`. 

As this is tutorial, so I just removed `global` key word.

In [11]:
%%bash
git config user.email "guangqiang.lu@accenture.com"
git config user.name "guangqiang.lu"

### 4. Create Repository

One thing to **highlight** is that full bellow steps should be done either with: `CMD` or `powershell` or `git bash` either is OK! Just type these keyword in `Search` in windows left corner!

#### 4. 1 Create a folder

Repository is a folder that contains your source code, you could create the folder manually or enter terminal with terminal , either is fine. Enter folder you need with `cd ***`, then create a folder like bellow: `mdkir learngit`, and enter the folder with `cd learngit`.

In [4]:
%%bash
mkdir learngit

In [5]:
%%bash
cd learngit

As I'm in notebook with `cd` not working, so I use Python to enter that folder.

In [6]:
import os

os.chdir('learngit')

#### 4.2 Init repository

Next we need to tell git that this folder should be tracked! How? with `git init`, then will create a hidden folder named: `.git`. 

If you couldn't see it, in file folder: click `View`, click `Hidden items`, then you could see that folder.

In [7]:
%%bash
git init

Initialized empty Git repository in /mnt/c/Users/guangqiiang.lu/Documents/lugq/github/machine-learning-beginner-jupyter-series/others/learngit/.git/


#### 4.3 Create share file

We could start create a shared file like txt or any other source code that needed to be tracked. You could create it any way you want, like right click with `create a new file`.

In [8]:
%%writefile readme.md
This is used for git tutorial!

Writing readme.md


#### 4.4 Put file into repository

As we have already created file and modified it, now we need to add it into repository(one thing to notice is that current this repository is located in local computer, we will make it into remote server later).

There will be 2 steps that we need to take, these 2 commands are used most times:
    1. `git add ***`
    2. `git commit -m ***`
    
`git commit -m ""` after `-m` is to provide short discription for the reason or what you have done.

In [9]:
%%bash
git add readme.md

In [12]:
%%bash
git commit -m "Add readme.md"

[master (root-commit) d8bf8f1] Add readme.md
 1 file changed, 1 insertion(+)
 create mode 100644 readme.md


After we commit, then we could get some information ` create mode 100644 readme.md` that git has already helped us create a track for file: `readme.md`, so if there is any change in future, git will help us remember it.

### 5. Git hands on

#### git status


We have already create one file, so let's change it to see what will happen in git with command: `git status`

In [13]:
%%writefile readme.md
This is used for git tutorial!

Add a new line for checking!

Overwriting readme.md


In [14]:
%%bash
git status

On branch master
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   readme.md

no changes added to commit (use "git add" and/or "git commit -a")


#### git diff

So we would like to know what we have chenged with this file, we could get the different part with: `git diff`. So we could see we are trying to add a new line to this file.

In [15]:
%%bash
git diff

diff --git a/readme.md b/readme.md
index d087c4a..b033001 100644
--- a/readme.md
+++ b/readme.md
@@ -1 +1,3 @@
 This is used for git tutorial!
+
+Add a new line for checking!


So that we could see that git tell us there is a change for file: readme.md, let's add our change into repository.

In [16]:
%%bash
git add readme.md

In [17]:
%%bash
git commit -m "Change readme.md"

[master 8622a90] Change readme.md
 1 file changed, 2 insertions(+)


We could try to confirm that our work tree is clean with command: git status.

In [18]:
%%bash
git status

On branch master
nothing to commit, working tree clean


#### git log

If we would try to get our work history, we could do it with: `git log`, we could see that there are 2 commit history with ID, name, date and file change.

In [19]:
%%bash
git log

commit 8622a901a77e8175d635181572e5a2ce90a188f3
Author: guangqiang.lu <guangqiang.lu@accenture.com>
Date:   Tue Jul 20 14:38:13 2021 +0800

    Change readme.md

commit d8bf8f12c01328ba48ff9143f36b9126091b0812
Author: guangqiang.lu <guangqiang.lu@accenture.com>
Date:   Tue Jul 20 14:36:05 2021 +0800

    Add readme.md


If there are too many for us, then we could add `--pretty=oneline` to get only main work with short des.

In [21]:
%%bash
git log --pretty=oneline

8622a901a77e8175d635181572e5a2ce90a188f3 Change readme.md
d8bf8f12c01328ba48ff9143f36b9126091b0812 Add readme.md


#### git reset

If we don't want our latest work to be moved into production, we could try to go to which version we need, for now there are 2 versions, supporse I want to go to first version, we could make it with: git reset 

In [26]:
%%bash
git reset --hard head^

HEAD is now at d8bf8f1 Add readme.md


In [27]:
%%bash
## check log, so that you could see that we are on first version.
git log --pretty=oneline

d8bf8f12c01328ba48ff9143f36b9126091b0812 Add readme.md


In [28]:
%%bash
# we could try to go to any version we like with that id, the first string of log with short string
git reset --hard d8bf8f12c01

HEAD is now at d8bf8f1 Add readme.md


In [29]:
%%bash
git log --pretty=oneline

d8bf8f12c01328ba48ff9143f36b9126091b0812 Add readme.md


#### git checkout

Sometimes we have already added our work into work folder with `git add`, then we don't need it to be added, how to remove? With git checkout.

In [58]:
%%writefile readme.md
This is used for git tutorial!

Add a new line for checking!

This is new added

Overwriting readme.md


In [59]:
%%bash
git status

On branch master
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   readme.md

no changes added to commit (use "git add" and/or "git commit -a")


In [60]:
%%bash
# remove added
git checkout -- readme.md

In [61]:
%%bash
# so that we see that change has been removed.
cat readme.md

This is used for git tutorial!


#### git remove

If we need to remove some files from our repository, use with `git remove`.

In [62]:
%%bash
# remove readme.md
rm readme.md

In [63]:
%%bash
# check current status that we have deleted file
git status

On branch master
Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	deleted:    readme.md

no changes added to commit (use "git add" and/or "git commit -a")


In [64]:
%%bash
# let's remove the files in repository
# If there are many changed files, we could try to use: git add . 
# to make full files.
git add .

In [65]:
%%bash
git commit -m "Remove readme.md"

[master a2c16d7] Remove readme.md
 1 file changed, 1 deletion(-)
 delete mode 100644 readme.md


### 6. Git branch

One important thing that need to be clear is that we could try to use different `branch` to seperate our `development` and `production` code, so they could be different for each environment, we could do change on `dev` branch, after we have tested, we could merge into `prd` branch.

Always we have **master** branch that should be the main branch for full circle and it should be in production env's code. So what you should work should create a new branch base on master and develop on that branch, after tested then merge it into master!

Let's try it out.

In [66]:
%%bash
# first try to create dev branch
# and now we are in dev branch.
git checkout -b dev

Switched to a new branch 'dev'


In [67]:
%%bash
# check which branch we are, * is where we are, 
git branch

* dev
  master


In [70]:
%%writefile readme.md
# make a file in dev branch
This is dev branch now!

Overwriting readme.md


In [71]:
%%bash
# check status
git status

On branch dev
Untracked files:
  (use "git add <file>..." to include in what will be committed)

	readme.md

nothing added to commit but untracked files present (use "git add" to track)


In [72]:
%%bash
# add it into work space
git add .

In [73]:
%%bash
git commit -m "Add readme.md into dev branch"

[dev 39484f0] Add readme.md into dev branch
 1 file changed, 2 insertions(+)
 create mode 100644 readme.md


In [75]:
%%bash
# check log
git log --pretty=oneline

39484f067648b6b634f561f68f20ab83b3a4b446 Add readme.md into dev branch
a2c16d70e63f8b097898f8c374dadf9bd84e174d Remove readme.md
d8bf8f12c01328ba48ff9143f36b9126091b0812 Add readme.md


In [76]:
%%bash
# checkout to master
git checkout master

Switched to branch 'master'


In [77]:
%%bash
# merge dev branch
git merge dev

Updating a2c16d7..39484f0
Fast-forward
 readme.md | 2 ++
 1 file changed, 2 insertions(+)
 create mode 100644 readme.md


In [80]:
%%bash
# check which branch we are, so we are in master
git branch

  dev
* master


In [79]:
%%bash
# so we see that we made our change into master
cat readme.md

# make a file in dev branch
This is dev branch now!


In [81]:
%%bash
# after we have merge our branch into master, we could delete current dev branch
git branch -d dev

Deleted branch dev (was 39484f0).


In [82]:
%%bash
git branch

* master


The last important thing is to notice is that when we try to merge some branch we may face that they are working on the same line of file, git won't know how to solve it, we need to manually solve these conflicts.

In [102]:
%%writefile readme.md
# make a file in master branch
This is dev branch now!

Overwriting readme.md


In [104]:
%%bash
git add .
git commit -m "change in master branch"

[master 1fad4ca] change in master branch
 1 file changed, 1 insertion(+), 1 deletion(-)


In [105]:
%%bash
# checkout a new branch
git checkout dev

Switched to branch 'dev'


In [108]:
%%writefile readme.md
# make a file in new dev branch
This is dev branch now!

Overwriting readme.md


In [109]:
%%bash
git add .
git commit -m "change in dev branch"

[dev f17f5d7] change in dev branch
 1 file changed, 1 insertion(+), 1 deletion(-)


In [110]:
%%bash
git checkout master

Switched to branch 'master'


So that we could see when we try to merge dev branch we get error that `CONFLICT (content): Merge conflict in readme.md`, so we need to manually solve this.

In [111]:
%%bash
git merge dev

Auto-merging readme.md
CONFLICT (content): Merge conflict in readme.md
Automatic merge failed; fix conflicts and then commit the result.


CalledProcessError: Command 'b'git merge dev\n'' returned non-zero exit status 1.

In [112]:
%%bash 
# now to check this file
cat readme.md

<<<<<<< HEAD
# make a file in master branch
# make a file in new dev branch
>>>>>>> dev
This is dev branch now!


What I mean manually sovle is that we need to open this file and `remove or change` some parts that we don't need for master branch and `save` it.

After I have manually changed it from local, then we could add it into our master again.

In [113]:
%%bash 
# now to check this file
cat readme.md

# make a file in master branch
This is dev branch now!


In [115]:
%%bash
git add .
git commit -m "Already solved conflict"

[master e79f0a8] Already solved conflict


In [116]:
%%bash
git log --pretty=oneline

e79f0a8e26faf13560e76488476be007b49f7ded Already solved conflict
f17f5d74724a70ac470e02e9794bf9612ce2bb3f change in dev branch
1fad4ca40a92518f3f2d013b4844d8b7756b8533 change in master branch
268f854554afe65a6d5648cdabb422cff48fff58 change in dev branch
6ddf8c022a24a88137bc1aaf05d88eb7257f814f change in dev branch
39484f067648b6b634f561f68f20ab83b3a4b446 Add readme.md into dev branch
a2c16d70e63f8b097898f8c374dadf9bd84e174d Remove readme.md
d8bf8f12c01328ba48ff9143f36b9126091b0812 Add readme.md


### Final words

This is a hands-on tutorial for **Git**, git is really a powerful tool that we should learn it and use it!

**Happy Learning Git**!